hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/TableServicePlanConflictDetection.java

```diff
@@ -0,0 +1,105 @@
+  public TableServicePlanConflictDetection(ConflictResolutionStrategy conflictResolutionStrategy, HoodieTableMetaClient metaClient, String lastKnownCompletionTime) {
+    this.conflictResolutionStrategy = conflictResolutionStrategy;
+    this.metaClient = metaClient;
+    this.lastKnownCompletionTime = lastKnownCompletionTime;
+  }
+
+  /**
+   * Checks if there are any potential conflicts with actions that have happened since the planner start instant.
+   * The compaction requires a check for conflicts with any commits for ingesting new data since the compaction plan is expected to contain all the log files for the slice it is compacting.
+   * It also requires checks for new plans for clustering or potentially another compaction on the timeline to avoid multiple services operating on the same file groups.
+   * @param compactionPlan the requested plan
+   * @param requestedInstantTime the requested instant time that will be used when persisting the plan to the timeline
+   * @return true if there is a conflict with the requested plan, false otherwise
+   */
+  public boolean hasConflict(HoodieCompactionPlan compactionPlan, String requestedInstantTime) {
+    ConcurrentOperation compactionOperation = new ConcurrentOperation(compactionPlan, requestedInstantTime);
+    return hasConflict(compactionOperation);
+  }
+
+  /**
+   * Checks if there are any potential conflicts with actions that have happened since the planner start instant.
+   * The clustering requires a check for conflicts with any commits for ingesting new data since the clustering plan will also contain log files,
+   * and all the log files for the slice it is clustering must be in the plan.
+   * It also requires checks for new plans for compaction or potentially another clustering on the timeline to avoid multiple services operating on the same file groups.
+   * @param clusteringPlan the requested plan
+   * @param requestedInstantTime the requested instant time that will be used when persisting the plan to the timeline
+   * @return true if there is a conflict with the requested plan, false otherwise
+   */
+  public boolean hasConflict(HoodieClusteringPlan clusteringPlan, String requestedInstantTime) {
+    ConcurrentOperation clusteringOperation = new ConcurrentOperation(clusteringPlan, requestedInstantTime);
+    return hasConflict(clusteringOperation);
+  }
+
+  private boolean hasConflict(ConcurrentOperation plannedOperation) {
+    return generateCandidateInstants().anyMatch(instant -> {
+      try {
+        ConcurrentOperation otherOperation = new ConcurrentOperation(instant, metaClient);
+        return conflictResolutionStrategy.hasConflict(plannedOperation, otherOperation);
+      } catch (IOException ex) {
+        throw new HoodieIOException("Unable to parse instant details for conflict detection", ex);
+      }
+    });
+  }
+
+  private Stream<HoodieInstant> generateCandidateInstants() {
+    Stream<HoodieInstant> commitsCompletedDuringPlanning = metaClient
```

Comment:
this is not necessarily completed right? why name it so? 
this is referring to pendingOrCompletedWriteCommits. 
or did you miss completed() filter in L 93, 94.

Reply:
I meant to add completed, will update and add a test

Reply:
There was a test but I had not refreshed the timeline so it was not working as expected

Reply:
yeah, I also did see that test and was puzzling.